File: src/main/java/ru/nsu/fit/team_project/serialization/CommandDeserializer.java
package ru.nsu.fit.team_project.serialization;

import com.google.gson.*;
import ru.nsu.fit.team_project.model.commands.Command;

import java.lang.reflect.Type;

public class CommandDeserializer implements JsonDeserializer<Command> {
    private static final String COMMAND_NAME_KEY = "commandName";
    private static final String COMMANDS_PACKAGE = "ru.nsu.fit.team_project.model.commands.";

    @Override
    public Command deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
        JsonObject jsonObject = json.getAsJsonObject();
        String className = jsonObject.get(COMMAND_NAME_KEY).getAsString();

        try {
            Class<?> clz = Class.forName(COMMANDS_PACKAGE + className);
            return context.deserialize(json, clz);
        } catch (ClassNotFoundException e) {
            throw new JsonParseException(e);
        }
    }
}

File: src/jvmMain/kotlin/data/serialization/CommandDeserializer.kt
package data.serialization

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import data.commands.entity.Command
import data.commands.entity.CommandType
import java.lang.reflect.Type

class CommandDeserializer : JsonDeserializer<Command> {

    private companion object {

        private const val COMMAND_TYPE_KEY = "commandType"
        private const val COMMANDS_PACKAGE = "data.commands.entity."
    }

    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Command {
        val jsonObject = json.asJsonObject
        val className = CommandType.valueOf(jsonObject.get(COMMAND_TYPE_KEY).asString)

        try {
            val clz = Class.forName(COMMANDS_PACKAGE + className)
            return context.deserialize(json, clz)
        } catch (e: ClassNotFoundException) {
            throw e
        }
    }
}